src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        inputConditon = {'RequestName': wflow['RequestName']}
+        try:
+            wmsReqInfo = self.wmstatsSvc.getFilteredActiveData(inputConditon, outputMask)
+            # self.logger.debug("wmsReqInfo: \n%s", pformat(wmsReqInfo))
+            if wmsReqInfo and (currentTime - wmsReqInfo[0]["RequestTransition"][-1]["UpdateTime"]) > threshold:
+                archDelayExpired = True
+        except Exception as ex:
+            msg = "General Exception while checking the time of the last request status transition for wflow: %s : %s"
+            self.logger.exception(msg, wflow['RequestName'], str(ex))
+        return archDelayExpired
+
+    def setArchDelayExpired(self, wflow):
+        """
+        A function to set the 'IsArchDelayExpired' flag
+        """
+        wflow['IsArchDelayExpired'] = self._checkArchDelayExpired(wflow)
+        return wflow
+
+    def _archive(self, wflow):
+        """
+        A function to make the actual request status transition during the
+        archival step.
+        :param wflow:       A MSRuleCleaner workflow representation
+        :return:            Bool
+        """
+        archiveStatus = False
+        try:
+            self.reqmgr2.updateRequestStatus(wflow['RequestName'], wflow['TargetStatus'])
+            archiveStatus = True
+        except Exception as ex:
+            msg = "General Exception while trying status transition to: %s for wflow: %s : %s"
+            self.logger.exception(msg, wflow['TargetStatus'], ['RequestName'], str(ex))
+        return archiveStatus
+
     def archive(self, wflow):
         """
         Move the workflow to the proper archived status after checking
         the full cleanup status
         :param  wflow:      A MSRuleCleaner workflow representation
-        :param  archStatus: Target status to transition after archival
         :return wflow:
         """
-        # NOTE: check allowed status transitions with:
-        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        # Make all the needed checks before trying to archive
         if not (wflow['IsClean'] or wflow['ForceArchive']):
             msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
             raise MSRuleCleanerArchival(msg)
+        elif not wflow['TargetStatus']:
+            msg = "Could not determine which archival status to target for: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+        elif not wflow['IsLogDBClean']:
+            if not self._cleanLogDB(wflow):
+                msg = "Could not clean logDB for: %s" % wflow['RequestName']
+                raise MSRuleCleanerArchival(msg)
+        elif not wflow['IsArchDelayExpired']:
+            msg = "Archival delay period has not yet expired for: %s." % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+        elif not self.msConfig['enableRealMode']:
```

Comment:
I guess you simply want to log a DRY-RUN thing here?

Reply:
Actually no. If we do not leave the function properly with raised exception here, then the workflow is going to get archived. 
But if you simply mean that the Error message should to be changed to "DRY-RUN  mode .... ", then I agree, that should be better.

Reply:
I see. However, we should distinguish the following two use cases:
1) workflow cannot be archived because there was an unexpected problem (e.g. wrong status, not all docs successfully cleaned, etc)
2) workflow is in an expected state (e.g. archival delay hasn't been met, service is in dry-run).

In addition to that, errors that are not really errors shouldn't be logged as error. What do you think?

Reply:
I totally agree. And the only way out of this I see if we put a second exception there and (re)name them both accordingly (Honestly speaking I do not like the name of the current exception either - it does not give any clue that this is actually an exception or an error of some kind)

Reply:
Done